stratum/hal/lib/barefoot/p4runtime_bfrt_translator.cc

```diff
@@ -0,0 +1,742 @@
+    const ::p4::v1::ReadRequest& request) {
+  absl::ReaderMutexLock l(&lock_);
+  if (!pipeline_require_translation_) {
+    return request;
+  }
+  ::p4::v1::ReadRequest translated_request(request);
+  for (::p4::v1::Entity& entity : *translated_request.mutable_entities()) {
+    ASSIGN_OR_RETURN(entity, TranslateEntity(entity, /*to_sdk=*/true));
+  }
+  return translated_request;
+}
+
+::util::StatusOr<::p4::v1::ReadResponse>
+P4RuntimeBfrtTranslator::TranslateReadResponse(
+    const ::p4::v1::ReadResponse& response) {
+  absl::ReaderMutexLock l(&lock_);
+  if (!pipeline_require_translation_) {
+    return response;
+  }
+  ::p4::v1::ReadResponse translated_response(response);
+  for (::p4::v1::Entity& entity : *translated_response.mutable_entities()) {
+    ASSIGN_OR_RETURN(entity, TranslateEntity(entity, /*to_sdk=*/false));
+  }
+  return translated_response;
+}
+
+bool P4RuntimeBfrtTranslator::ReadResponseWriterWrapper::Write(
+    const ::p4::v1::ReadResponse& msg) {
+  auto status = p4runtime_bfrt_translator_->TranslateReadResponse(msg);
+  return status.ok() && writer_->Write(status.ConsumeValueOrDie());
+}
+
+bool P4RuntimeBfrtTranslator::StreamMessageResponseWriterWrapper::Write(
+    const ::p4::v1::StreamMessageResponse& msg) {
+  auto status = p4runtime_bfrt_translator_->TranslateStreamMessageResponse(msg);
+  return status.ok() && writer_->Write(status.ConsumeValueOrDie());
+}
+
+::util::Status P4RuntimeBfrtTranslator::PushChassisConfig(
+    const ChassisConfig& config, uint64 node_id) {
+  ::absl::WriterMutexLock l(&lock_);
+  // Port mapping for P4Runtime translation.
+  singleton_port_to_sdk_port_.clear();
+  sdk_port_to_singleton_port_.clear();
+  // Initialize with special ports.
+  singleton_port_to_sdk_port_[kSdnUnspecifiedPortId] = 0;
+  ASSIGN_OR_RETURN(const auto& cpu_sdk_port,
+                   bf_sde_interface_->GetPcieCpuPort(device_id_));
+  singleton_port_to_sdk_port_[kSdnCpuPortId] = cpu_sdk_port;
+  sdk_port_to_singleton_port_[cpu_sdk_port] = kSdnCpuPortId;
+  for (int pipe = 0; pipe < 4; pipe++) {
+    uint32 sdk_port = kTnaRecirculationPortBase | (pipe << 7);
+    uint32 sdn_port = kSdnTnaRecirculationPortBase + pipe;
+    singleton_port_to_sdk_port_[sdn_port] = sdk_port;
+    sdk_port_to_singleton_port_[sdk_port] = sdn_port;
+  }
+
+  for (const auto& singleton_port : config.singleton_ports()) {
+    uint32 singleton_port_id = singleton_port.id();
+    PortKey singleton_port_key(singleton_port.slot(), singleton_port.port(),
+                               singleton_port.channel());
+    ASSIGN_OR_RETURN(uint32 sdk_port_id,
+                     bf_sde_interface_->GetPortIdFromPortKey(
+                         device_id_, singleton_port_key));
+    singleton_port_to_sdk_port_[singleton_port_id] = sdk_port_id;
+    sdk_port_to_singleton_port_[sdk_port_id] = singleton_port_id;
+  }
+
+  return ::util::OkStatus();
+}
+
+::util::Status P4RuntimeBfrtTranslator::PushForwardingPipelineConfig(
```

Comment:
> Note that it may be infeasible to translate the value-mask pair for ternary matches: LPM, TERNARY or RANGE match kinds. The P4Runtime server may require that for these match kinds the port match be either de facto “exact” (0xffffffff mask for TERNARY, prefix-length of 32 for LPM, or same low and high bounds for RANGE) or “don't care”. 

~~Are we checking this somewhere?~~ ok, we do in `TranslateTableEntry`. Can we move this check into `PushForwardingPipelineConfig`? Much earlier.

Reply:
Sorry, I think I don't understand your question.
We need to check the value/mask/prefix_len/range of match fields for each table entry when we write table entries. That's the reason I put check in the `TranslateTableEntry`

Reply:
my bad, I confused two things here. The check for `0xffffffff`/prefix len/don't care can, of course, only happen at flow push time.